### PR TITLE
[RFC] Proposed API for options cache

### PR DESCRIPTION
**Pull request by nicolasvasilache**
Options cache is used only in the autotuner.
Upon further inspection and rewriting of the autotuner logic I found the cache implementation was convoluted by virtue of trying to do too much refactoring with the CRTP-based cache used for compilation.
This is a simpler (claim just my own :) )  implementation that uses the same key/value representation in both proto and memory.
A lightweight wrapper with conversion to/from proto is provided as usual because the proto API doesn't hold a candle to STL.

As far as the proto changes go please only look at the part related to OptionsCache, the rest is unrelated.
I can try to split it off if it is too hard to parse for people.

A subsequent PR may propose to move to proto3 and use their map but it is too much of a change for now.

### CODE REVIEW

File: tc/autotuner/options_cache.h [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fillFromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromProtobuf(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void dumpCacheToProtobuf(const std::string& filename);
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/autotuner/options_cache.h]: > Can protobuf be serialized to some other type of stream instead of a file, network stream, string

I would hold that off until we do it consistently for the compilation caches (hint, at a later date).

File: tc/autotuner/options_cache.h [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      const std::vector<TensorInfo>& outputs,
+      const std::string& deviceStr,
+      const typename Backend::MappingOptionsType& options,
+      Duration duration);
+
+  /// Returns the best mapping options for a particular TC/inputs/outputs/device
+  /// Note that the result may be empty (in particular if problem size is
+  /// small and pruning threshold is too high for the problem size)
+  /// \returns an optional mapping option
+  llvm::Optional<typename Backend::MappingOptionsType> getBestOptions(
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/autotuner/options_cache.h]: Yeah .. sigh we've been ping-ponging between ```unique_ptr```, ```llvm::optional``` and our own ```Optional``` impl.

File: tc/autotuner/options_cache.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+  //// Calls the underlying unordered_multimap::size
+  size_t size() const;
+
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fillFromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/autotuner/options_cache.h]: Because which one do you keep by default? original or old?
What behavior is preferred in case of duplication? skip or overwrite?
Duplicates themselves are not an issue, clear the cache if you don't want them.
As a side note, use cases of ```OptionsCache:::loadFromFile``` are on cleared caches.

File: tc/proto/compcache.proto [hunk 4]
```diff
@@ -16,57 +16,72 @@ message TensorInfoProto {
...
+  repeated uint64 recorded_runtimes = 2;
 }
 
+message CpuOptionsCacheProto {
+  repeated OptionsCacheKeyProto keys = 1;
+  repeated CpuOptionsCacheValueProto values = 2;
+}
+
+/******************************************************************************/
+/* CUDA-specific cache                                                        */
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/proto/compcache.proto]: it is, moving

File: tc/proto/compcache.proto [hunk 5]
```diff
@@ -16,57 +16,72 @@ message TensorInfoProto {
...
 }
 
-message OptionsCacheEntryProto {
-  required string id = 1;
-  repeated TensorInfoProto inputs = 2;
-  repeated TensorInfoProto outputs = 3;
-  required string device_str = 4;
-  required string git_version = 5;
+message CudaOptionsCacheProto {
+  repeated OptionsCacheKeyProto keys = 1;
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/proto/compcache.proto]: All these things should go away and we should use ```proto3``` with a simple map.

> the key and the value merged in one message and don't have this problem

I'd claim they have many other problems. Look in particular at the implementation of CudaCompilationCache which is currently distributed across [1](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/compilation_cache.h), [2](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/compilation_cache.cc), [3](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache-inl.h), [4](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache.cc), [5](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache.h) with CRTP fun.

Some of the complexity comes from specializing this data structure. But many problems come from trying to be too generic too early and all these templated searchKernel functions.

In this refactoring (and this OptionsCache RFC) I am going back to basics, backend templated multimap which matches the semantics of what an OptionsCache provides. We'll tend to problems when we hit them. All the rest is premature optimization which is now blowing to our faces when we try to be truly backend-agnostic.

One part of me really feel like ripping apart all these compilation caches as premature optimizations because in the new API we have the tuner with its cache (proposed here) and that's really the expensive part. The rest frankly could be tore down for now and redone from scratch. I already tore down the manual caches because in practice they are useless for now.

What would you think of tearing the CompilationCache apart for now?
Once autotuned, one would JIT-recompile anyway (and can reuse the same options for different input sizes) and can hold on to the TcExecutor.

cc @ttheodor

File: tc/proto/compcache.proto [hunk 6]
```diff
@@ -16,57 +16,72 @@ message TensorInfoProto {
...
+message CpuOptionsCacheProto {
+  repeated OptionsCacheKeyProto keys = 1;
+  repeated CpuOptionsCacheValueProto values = 2;
+}
+
+/******************************************************************************/
+/* CUDA-specific cache                                                        */
+/******************************************************************************/
+message CudaCacheEntryProto {
+  required string id = 1;
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/proto/compcache.proto]: Yes but the contorsions required to go from a vector of entries to a key value pair structure and propagating all this through the codebase are ***painful***, so I'll leave this for later.
All these things should go away and we should use ```proto3``` with a simple map; this is all technical debt that takes a lot of time to clean.

File: tc/proto/compcache.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply tc/proto/compcache.proto]: And here is what a commit that drops compilation caches [looks like](https://github.com/nicolasvasilache/TensorComprehensions/commit/0f0d392012c6dfc6da75132935a25abfddc948bb) in the new API.

Compiles, tests pass etc.

[the remark this answers is not in the document]
Comment (ttheodor) [reply tc/proto/compcache.proto]: @nicolasvasilache  it doesn't sound like a bad idea. The cuda caches were used by the autotuner to avoid recompilations, however now that execution engine is dead (or will be soon) we can use the executors directly instead of relying on an underlying cache.

[the remark this answers is not in the document]
Comment (ftynse) [reply tc/proto/compcache.proto]: I discussed with @math-fehr and it seems that compilation caches create overhead when we want to use cooperative groups (launching cooperative kernels uses a different call and adds extra overhead so we don't want to use it always).

So I would second the decision to remove the compilation caches.

File: tc/autotuner/options_cache.h [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  /// Loads in place from proto file. Calls fromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromFile(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void storeCacheToFile(const std::string& filename) const;
+
+  /// Saves a new runtime.
+  /// If the key does not exist, a new entry is inserted.
+  /// If the key exists, a linear search is performed on options.
```
Comment (skimo-openhub) [inline tc/autotuner/options_cache.h]: This describes _how_ the function works (linear search), rather than _what_ it does.
Is this necessary here?

File: tc/autotuner/options_cache.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (skimo-openhub) [reply tc/autotuner/options_cache.h]: How about simply removing `getBestOptions` and letting the user call `getTopKOptions` with `K` set to 1 instead?

**Comment by nicolasvasilache** [comment]
served its purpose
